teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+| `TaskRun`, `PipelineRun` | `Unknown` | `Running` | `running` | `started`, `running` or None | `started` or None | If the `started` event was not sent yet, it is sent. If the `running` event was not sent yet, it is sent. If both events were sent already, the `running` event is sent only if there was a change in the `Condition` compared to the last `running` event sent |
+| `TaskRun`, `PipelineRun` | `Unknown` | Any but `Running` | `unknown` | `started`, `unknown` or None | `started` or None | If the `started` event was not sent yet, it is sent. If the `unknown` event was not sent yet, it is sent. If both events were sent already, the `unknown` event is sent only if there was a change in the `Condition` compared to the last `unknown` event sent |
+| `CustomRun` | `Unknown` | Any | `running` | `started`, `running` | `started` or None | If the `started` event was not sent yet, it is sent. We cannot make assumptions about how `Reason` and `Message` are used by the custom run controller, so always send a `running` event |
+| Any         | `Succeed` | Any | `successful` | `successful` | `finished` | The `finished` CDEvent include the `Condition` in the `outcome` field |
+| Any         | `Failed` | Any | `failed` | `failed` | `finished` | The `finished` CDEvent include the `Condition` in the `outcome` field |
+
+Since the logic for sending events depends on the content of the cache, it's important
+for the cache to be updated **in order**.
+
+### Notes and Caveats
+
+<!--
+(optional)
+
+Go in to as much detail as necessary here.
+- What are the caveats to the proposal?
+- What are some important details that didn't come across above?
+- What are the core concepts and how do they relate?
+-->
+
+
+## Design Details
+
+Example configuration:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-events
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+data:
+  # formats contains a comma separated list of event formats to be used
+  # the only format supported today is "tektonv1". An empty string is not
+  # a valid configuration. To disable events, do not specify the sink.
+  formats: "tektonv1,cdevents"
+
+  # sink contains the event sink to be used for TaskRun, PipelineRun and
+  # CustomRun. If no sink is specified, no CloudEvent is generated.
+  # This setting supercedes the "default-cloud-events-sink" from the
+  # "config-defaults" config map
+  sink: "https://events.sink/cdevents"
+```
+
+Example CDEvent:
+
+```yaml
+{
+  "context": {
+    "version": "0.3.0",
+    "id": "271069a8-fc18-44f1-b38f-9d70a1695819",
+    "source": "/apis/tekton.dev/v1/namespaces/default/taskruns/print-date-q4hv8",
+    "type": "dev.cdevents.pipelinerun.started.0.1.1",
```

Comment:
thanks @afrittoli for adding an example, please help me understand 🙏 

`source` represents `taskrun` but the `type` is `pipelinerun`? or am I reading this incorrectly?

Reply:
Thanks @pritidesai, nice catch, my example is incorrect, I will fix it.